catroid/src/main/java/org/catrobat/catroid/ProjectManager.java

```diff
@@ -99,105 +99,30 @@ public void loadProject(File projectDir, Context context) throws ProjectExceptio
-			project.setCatrobatLanguageVersion(0.95f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.95f) {
-			project.setCatrobatLanguageVersion(0.96f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.96f) {
-			project.setCatrobatLanguageVersion(0.97f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.97f) {
-			project.setCatrobatLanguageVersion(0.98f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.98f) {
-			project.setCatrobatLanguageVersion(0.99f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.99f) {
-			project.setCatrobatLanguageVersion(0.991f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.991f) {
-			project.setCatrobatLanguageVersion(0.992f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.992f) {
-			updateCollisionFormulasTo993(project);
-			project.setCatrobatLanguageVersion(0.993f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.993f) {
-			updateSetPenColorFormulasTo994(project);
-			project.setCatrobatLanguageVersion(0.994f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.994f) {
-			updateArduinoValuesTo995(project);
-			project.setCatrobatLanguageVersion(0.995f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.995f) {
-			updateCollisionScriptsTo996(project);
-			project.setCatrobatLanguageVersion(0.996f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.996f) {
-			project.setCatrobatLanguageVersion(0.997f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.997f) {
-			project.setCatrobatLanguageVersion(0.998f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.998f) {
-			project.setCatrobatLanguageVersion(0.999f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.999f) {
-			makeShallowCopiesDeepAgain(project);
-			project.setCatrobatLanguageVersion(0.9991f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.9991f) {
-			project.setCatrobatLanguageVersion(0.9992f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.9992f) {
-			project.setCatrobatLanguageVersion(0.9993f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.9993f) {
-			updateScriptsToTreeStructure(project);
-			project.setCatrobatLanguageVersion(0.9994f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.9994f) {
-			project.setCatrobatLanguageVersion(0.9995f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.9995f) {
-			project.setCatrobatLanguageVersion(0.9996f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.9996f) {
-			project.setCatrobatLanguageVersion(0.9997f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.9997f) {
-			project.setCatrobatLanguageVersion(0.9998f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.9998f) {
-			project.setCatrobatLanguageVersion(0.9999f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.9999f) {
-			project.setCatrobatLanguageVersion(0.99991f);
-		}
-		if (project.getCatrobatLanguageVersion() == 0.99991f) {
+		if (checkValidLanguageVersion(project.getCatrobatLanguageVersion())) {
```

Comment:
if youd do the check here and return right away if invalid you save one level of intentation and have the parts belonging together in one place:

`if (!checkValidLanguageVersion(project.getCatrobatLanguageVersion())) {
			restorePreviousProject(previousProject);
			throw new CompatibilityProjectException(context.getString(R.string.error_project_compatibility));
}`

Reply:
you could even invert the method to already say `isInvalidLanguageVersion()` then this approach would look even cleaner (look at the snippet below). besides you could even leave out the `else` because the exception stops method execution anyway. so that's two levels of indentation and felt cyclomatic complexity. 

if you go this way, just don't forget to invert the return value in the method FKA `checkValidLanguageVersion`.

```
if (isInvalidLanguageVersion(project.getCatrobatLanguageVersion()) {
    restorePreviousProject(previousProject);
    throw new CompatibilityProjectException(context.getString(R.string.error_project_compatibility));
}

if (project.getCatrobatLanguageVersion() <= 0.91f) {
   project.setScreenMode(ScreenModes.STRETCH);
}
if (project.getCatrobatLanguageVersion() <= 0.992f) {
    updateCollisionFormulasTo993(project);
}
if (project.getCatrobatLanguageVersion() <= 0.993f) {
    updateSetPenColorFormulasTo994(project);
}
if (project.getCatrobatLanguageVersion() <= 0.994f) {
    updateArduinoValuesTo995(project);
}
if (project.getCatrobatLanguageVersion() <= 0.995f) {
    updateCollisionScriptsTo996(project);
}
if (project.getCatrobatLanguageVersion() <= 0.999f) {
    makeShallowCopiesDeepAgain(project);
}
if (project.getCatrobatLanguageVersion() <= 0.9993f) {
    updateScriptsToTreeStructure(project);
}			
project.setCatrobatLanguageVersion(CURRENT_CATROBAT_LANGUAGE_VERSION);